### PR TITLE
driver: adc: infineon: Adding ADC driver

### PR DESCRIPTION
**Pull request by billwatersiii**
- This includes the driver, test app, and sample app
- Only the boards\arm\cy8cproto_062_4343w board is supported for now

### CODE REVIEW

File: drivers/adc/adc_ifx_cat1.c [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+					      bool repeat_sampling)
+{
+	struct ifx_cat1_adc_data *data = CONTAINER_OF(ctx, struct ifx_cat1_adc_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static void adc_context_on_complete(struct adc_context *ctx, int status)
```
Comment (npal-cy) [inline drivers/adc/adc_ifx_cat1.c:102]: please remove if it does not used.

Comment (billwatersiii) [reply drivers/adc/adc_ifx_cat1.c]: Done

**Review by ifyall — DISMISSED** [review]

**Review by npal-cy — DISMISSED** [review]

**Comment by billwatersiii** [comment]
> 

The commit with the changes to those files is no longer part of this PR

**Comment by billwatersiii** [comment]
> Hi Bill, could you please look at:
> Conflicting files
> drivers/sensor/CMakeLists.txt
> drivers/sensor/Kconfig

The commit with the changes to those files is no longer part of this PR

File: drivers/adc/adc_ifx_cat1.c [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+/* Default Counter interrupt priority */
+#if !defined(IFX_CAT1_ADC_INTERRUPT_PRIORITY)
+	#define IFX_CAT1_ADC_INTERRUPT_PRIORITY (3u)
+#endif
+
+#define ADC_CAT1_EVENTS_MASK (CYHAL_ADC_EOS | CYHAL_ADC_ASYNC_READ_COMPLETE)
+
+#define ACQUISITION_TIME_NS (1000u)
+#define ADC_RESOLUTION      (12u)
+#define REF_INTERNAL_MV     (1200u)
```
Comment (npal-cy) [inline drivers/adc/adc_ifx_cat1.c:39]: please rename:
ACQUISITION_TIME_NS ->ADC_CAT1_ACQUISITION_TIME_NS
ADC_RESOLUTION -> ADC_CAT1_RESOLUTION
ACQUISITION_TIME_NS -> ADC_CAT1_ACQUISITION_TIME_NS

File: drivers/adc/adc_ifx_cat1.c [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+LOG_MODULE_REGISTER(ifx_cat1_adc, CONFIG_ADC_LOG_LEVEL);
+
+#if defined(PASS_SARMUX_PADS0_PORT)
+	#define _ADCSAR_PORT PASS_SARMUX_PADS0_PORT
+#else
+	#error The selected ADC device is not supported...
+#endif
+
+/* Default Counter interrupt priority */
+#if !defined(IFX_CAT1_ADC_INTERRUPT_PRIORITY)
```
Comment (npal-cy) [inline drivers/adc/adc_ifx_cat1.c:31]: remove this, now it handled via ifx_cat1_adc_config/irq_priority

File: drivers/adc/adc_ifx_cat1.c [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ifx_cat1_adc, CONFIG_ADC_LOG_LEVEL);
+
+#if defined(PASS_SARMUX_PADS0_PORT)
+	#define _ADCSAR_PORT PASS_SARMUX_PADS0_PORT
+#else
+	#error The selected ADC device is not supported...
```
Comment (npal-cy) [inline drivers/adc/adc_ifx_cat1.c:27]: change to "#error The selected device is not supported ADC ..."

**Review by npal-cy — CHANGES_REQUESTED** [review]

**Review by npal-cy — DISMISSED** [review]

File: drivers/adc/adc_ifx_cat1.c [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+	};								       \
+									       \
+	static const struct ifx_cat1_adc_config adc_cat1_cfg_##n = {	       \
+		.irq_priority = DT_INST_IRQ(n, priority),		       \
+	};								       \
+									       \
+	DEVICE_DT_INST_DEFINE(n, ifx_cat1_adc_init,			       \
+			      NULL, &ifx_cat1_adc_data##n,		       \
+			      &adc_cat1_cfg_##n,			       \
+			      POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE, \
```
Comment (anangl) [inline drivers/adc/adc_ifx_cat1.c:268]:
```suggestion
			      POST_KERNEL, CONFIG_ADC_INIT_PRIORITY,	       \
```
See b0cdef3c25c2ac92627f158700efdb9203ecb51a.

File: drivers/adc/adc_ifx_cat1.c [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+		/* Minimum acquisition time set to 1us */
+		.min_acquisition_ns = ADC_CAT1_ACQUISITION_TIME_NS,
+		/* Sample channel when ADC performs a scan */
+		.enabled = true
+	};
+
+	/* Initialize a channel and configure it to scan the input pin(s). */
+	cyhal_adc_channel_free(&data->adc_chan_obj[channel_cfg->channel_id]);
+	result = cyhal_adc_channel_init_diff(&data->adc_chan_obj[channel_cfg->channel_id],
+					     &data->adc_obj, vplus, vminus, &channel_config);
```
Comment (anangl) [inline drivers/adc/adc_ifx_cat1.c:120]: Why `gain`, `reference`, and `acquisition_time` from `channel_cfg` are ignored here? The driver should return `-EINVAL` if it cannot set the requested configuration.

File: drivers/adc/adc_ifx_cat1.c (hunk 5 above)
Comment (billwatersiii) [reply drivers/adc/adc_ifx_cat1.c]: Done.  Thanks for pointing this out.

File: drivers/adc/adc_ifx_cat1.c (hunk 6 above)
Comment (billwatersiii) [reply drivers/adc/adc_ifx_cat1.c]: Done

**Review by anangl — DISMISSED** [review]

**Review by npal-cy — DISMISSED** [review]

**Comment by carlescufi** [comment]
@billwatersiii please rebase

**Review by anangl — APPROVED** [review]

**Review by npal-cy — APPROVED** [review]

**Review by ifyall — APPROVED** [review]

**Pull request merged**